# piecharts for ever detected proteins from Cys and Lys experiments

# PANTHER analysis of protein class

"""
Created on 11.23.19
for Project MAP
@author: mfpfox
"""

citation
```
PANTHER version 14: more genomes, a new PANTHER GO-slim and improvements in enrichment analysis tools.
Huaiyu Mi, Anushya Muruganujan, Dustin Ebert, Xiaosong Huang and Paul D. Thomas
Nucl. Acids Res. (2019) doi: 10.1093/nar/gky1038
```

In [1]:
import os
import sys
import pandas as pd

sys.path.append("/Users/mariapalafox/Desktop/Toolbox")
from all_funx import *

from IPython.display import display, HTML
pd.set_option('display.max_columns', None)
pd.options.display.max_seq_items = 2000
display(HTML("<style>.container {width:90% !important;}</style>"))

---

# version 1 pie n=3963 (all uniprot IDs from 2012->2018 mapping)
- before filtering for IDs contained in all ensembl releases

**below function in toolbox "Pmap_pie.py"**

In [2]:
os.chdir("/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CYS_LYS_2012_to2018UKBpositions/REFERENCE_2018_mappedCK/SUBSETS/")

In [3]:
# ukb file with be one annotated
everC = "CYS2018_everlabeled_gene_valueCounts_2915.csv"
everK = "LYS2018_everlabeled_gene_valueCounts_2623.csv"
uni = "countsfromfastaCCDS_filter_counts_all_AA.csv"
shared3953 = "UKBID_3953set.csv"

cys = read_pd(everC)
lys = read_pd(everK)
ukb = read_pd(uni)
dropUnnamed(ukb)

shared = read_pd(shared3953)

print("ALL 3963 uniprot IDs: ")
print("total of all lysines detected: ", lys.counts.sum())
print("total of all cysteine detected: ", cys.counts.sum())

# testing main()
Cls = cys.ID.tolist()
Kls = lys.ID.tolist()

ukbC = addcolumnconditional(Cls, ukb, 'ID', 'ever_C_ID')
ukbCK = addcolumnconditional(Kls, ukbC, 'ID', 'ever_K_ID')
print("shape ukbCK labeled = ", ukbCK.shape)

# only C detected proteins
Conly = ukbCK[(ukbCK['ever_C_ID'] == "True") & (ukbCK['ever_K_ID'] == "False")]
print("TOTAL protiens with only Cys detected: ")
print(Conly.shape)
print("TOTAL Cys in the proteins: ", Conly.C.sum())
print()

# only K detected proteins
Konly = ukbCK[(ukbCK['ever_C_ID'] == "False") & (ukbCK['ever_K_ID'] == "True")]
print("TOTAL protiens with only Lysine detected: ")
print(Konly.shape)
print("TOTAL Lysine in the proteins: ", Konly.K.sum())
print()

# both C K detecte proteins
bothck = ukbCK[(ukbCK['ever_C_ID'] == "True") & (ukbCK['ever_K_ID'] == "True")]
print("TOTAL protiens with both CK detected: ")
print(bothck.shape)
print("TOTAL Lysine in the proteins: ", bothck.K.sum())
print("TOTAL Cysteine in the proteins: ", bothck.C.sum())
print()

# all C IDs
Call = ukbCK[ukbCK['ever_C_ID'] == "True"]
print("TOTAL protiens with all C detected: ")
print(Call.shape)
print("TOTAL C in the proteins: ", Call.C.sum())
print()

Kall = ukbCK[ukbCK['ever_K_ID'] == "True"]
print("TOTAL protiens with all K detected: ")
print(Kall.shape)
print("TOTAL K in the proteins: ", Kall.K.sum())

ALL 3963 uniprot IDs: 
total of all lysines detected:  9042
total of all cysteine detected:  6234
checking labeled column values: 
  ever_C_ID  Count
0     False  15517
1      True   2915

checking labeled column values: 
  ever_K_ID  Count
0     False  15809
1      True   2623

shape ukbCK labeled =  (18432, 30)
TOTAL protiens with only Cys detected: 
(1340, 30)
TOTAL Cys in the proteins:  19812

TOTAL protiens with only Lysine detected: 
(1048, 30)
TOTAL Lysine in the proteins:  34910

TOTAL protiens with both CK detected: 
(1575, 30)
TOTAL Lysine in the proteins:  61917
TOTAL Cysteine in the proteins:  14944

TOTAL protiens with all C detected: 
(2915, 30)
TOTAL C in the proteins:  34756

TOTAL protiens with all K detected: 
(2623, 30)
TOTAL K in the proteins:  96827


In [8]:
cys.head(3)

ID  counts  final_refID
0  P49327      25         True
1  P21333      25         True
2  Q15149      22         True

In [9]:
lys.head(3)

ID  counts  final_refID
0  Q09666     105         True
1  P21333      50         True
2  O75369      47         True

---
---
---


# version 2 PIE chart with 3,953 IDs shared across all ensembl releases

In [5]:
# 3953 ID version 
fil = shared.ID.tolist()
print("length of list: ", len(fil))
print("length of set: ", len(set(fil)))
print()
lysdrop = addcolumnconditionalDrop(fil, lys, 'ID', 'set3953_ID')
cysdrop = addcolumnconditionalDrop(fil, cys, 'ID', 'set3953_ID')
print("shape lys drop  = ", lysdrop.shape)
print("shape cys drop  = ", cysdrop.shape)

length of list:  3953
length of set:  3953

  set3953_ID  Count
0       True   2618
dropping false
df shape post drop:  (2618, 4)
  set3953_ID  Count
0       True   2909
dropping false
df shape post drop:  (2909, 4)
shape lys drop  =  (2618, 4)
shape cys drop  =  (2909, 4)


In [6]:
idsimple = pd.DataFrame(fil)
idsimple.columns = ['ID']
idsimple.shape

(3953, 1)

In [7]:
# saving for later
#idsimple.to_csv("/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/id3953.csv", index=False)

In [8]:
print("ALL 3953 uniprot IDs: ")
print("total of all lysines detected: ", lysdrop.counts.sum())
print("total of all cysteines detected: ", cysdrop.counts.sum())

# change df name
Cls = cysdrop.ID.tolist()
Kls = lysdrop.ID.tolist()

ukbC = addcolumnconditional(Cls, ukb, 'ID', 'ever_C_ID')
ukbCK = addcolumnconditional(Kls, ukbC, 'ID', 'ever_K_ID')
print("shape ukbCK labeled = ", ukbCK.shape)

# only C detected proteins
Conly = ukbCK[(ukbCK['ever_C_ID'] == "True") & (ukbCK['ever_K_ID'] == "False")]
print("TOTAL protiens with only Cys detected: ")
print(Conly.shape)
print("TOTAL Cys in the proteins: ", Conly.C.sum())
print()

# only K detected proteins
Konly = ukbCK[(ukbCK['ever_C_ID'] == "False") & (ukbCK['ever_K_ID'] == "True")]
print("TOTAL protiens with only Lysine detected: ")
print(Konly.shape)
print("TOTAL Lysine in the proteins: ", Konly.K.sum())
print()

# both C K detecte proteins
bothck = ukbCK[(ukbCK['ever_C_ID'] == "True") & (ukbCK['ever_K_ID'] == "True")]
print("TOTAL protiens with both CK detected: ")
print(bothck.shape)
print("TOTAL Lysine in the proteins: ", bothck.K.sum())
print("TOTAL Cysteine in the proteins: ", bothck.C.sum())
print()

# all C IDs
Call = ukbCK[ukbCK['ever_C_ID'] == "True"]
print("TOTAL protiens with all C detected: ")
print(Call.shape)
print("TOTAL C in the proteins: ", Call.C.sum())
print()

Kall = ukbCK[ukbCK['ever_K_ID'] == "True"]
print("TOTAL protiens with all K detected: ")
print(Kall.shape)
print("TOTAL K in the proteins: ", Kall.K.sum())


ALL 3953 uniprot IDs: 
total of all lysines detected:  9034
total of all cysteines detected:  6225
checking labeled column values: 
  ever_C_ID  Count
0     False  15523
1      True   2909

checking labeled column values: 
  ever_K_ID  Count
0     False  15814
1      True   2618

shape ukbCK labeled =  (18432, 30)
TOTAL protiens with only Cys detected: 
(1335, 30)
TOTAL Cys in the proteins:  19768

TOTAL protiens with only Lysine detected: 
(1044, 30)
TOTAL Lysine in the proteins:  34814

TOTAL protiens with both CK detected: 
(1574, 30)
TOTAL Lysine in the proteins:  61905
TOTAL Cysteine in the proteins:  14940

TOTAL protiens with all C detected: 
(2909, 30)
TOTAL C in the proteins:  34708

TOTAL protiens with all K detected: 
(2618, 30)
TOTAL K in the proteins:  96719


---
---
---


---
---
---


# PANTHER PROTEIN CLASSIFICATION

version = PANTHER14.1 is generated from the **2018_04 release of ReferenceProteome dataset**

<http://www.pantherdb.org/chart/pantherChart.jsp?listType=1&filterLevel=1&type=5&chartType=2&save=yes&basketItems=all&zoom=1&trackingId=626614EA3D2EAB0581A6BA3EDA4F01F5>

## genes in protein classes not mutually exclusive
- used PANTHER molecular function, biological process, and protein class annotation
    - protein class ontology adapted fromm PANTHER/X molecular function ontology

### how many specific groups do i have from PANTHER automatic annotation?
- similar molecular function groups = 
- similar biological process groups = 
- protein class groups = 




### unable to map 14 IDs out of (3963 and 3953 same unfound IDs):

```
A6ZKI3
O60656
P42167
P42773
P50135
P60896
P61513
Q13765
Q14116 
Q15121
Q99470
Q9GZP8
Q9NRS6
Q9Y3B4
```

#### subsets of missed IDs - using 3963 set
##### from cys ever n = 2915
```
# UNIQUE TO C EVER LIST (1)
Q14116
# IN LYS MISMAP LIST (5)
P50135
Q15121
Q9Y3B4
P42167
P61513
```
##### from lys ever n = 2623
```
# UNIQUE TO K EVER LIST (8)
Q13765
Q99470
A6ZKI3
Q9GZP8
P60896
P42773
Q9NRS6
O60656
# IN CYS MISMAP LIST (5)
P61513
P42167
Q9Y3B4
Q15121
P50135
```

**6 + 8 = 14**


| Not mapped in both C and K detected ID list | protein primary name from ukb website |
|------------------------------------|-------------------------------------|
| P61513 | 60S ribosomal protein L37a, gene(RPL37A) |
| P42167 | Lamina-associated polypeptide 2, isoforms beta/gamma, gene(TMPO) |
| Q9Y3B4 | Splicing factor 3B subunit 6, gene(SF3B6) |
| Q15121 | Astrocytic phosphoprotein PEA-15, gene(PEA15) |
| P50135 | Histamine N-methyltransferase, gene(HNMT) |


## Ontologies PANTHER™ GO slim (version 14.1, based on GO release 2018-07-03, released 2019-03-12)
- 3039 total terms
- 2004 biological process terms
- 512 cellular component terms
- 523 molecular function terms
- PANTHER™ Protein Class (version 14.1, released 2019-03-12)
    - 256 total terms
    
    
Accessed : PANTHER™ Data at a glance on 12.15.19 (above info copied from website)

# MISCAST {Iqbal,2019} protein class assignment rules from methods:
1. Ensembl family description
2. molecular funcx/biological process annotation previous collected from PANTHER or 
3. molecular funcx/biological process annotation available in UniProt as definded by teh gene ontology consortium

In [25]:
# miscast 24 protein classes for 1330 genes
miscast = pd.read_csv("/Users/mariapalafox/Desktop/myMAP_notes/MISCAST_BROAD/miscast_24Protein_classes.csv")
miscast

Protein Class  Gene count  Structure count  \
0                   All proteins        1330            14270   
1        Calcium-Binding Protein          35              311   
2         Cell Adhesion Molecule          40              298   
3          Cell Junction Protein          24              128   
4                      Chaperone          27              137   
5           Cytoskeletal Protein          83              635   
6       Defense/Immunity protein          57             1069   
7               Enzyme Modulator         119             1280   
8   Extracellular Matrix Protein          41              309   
9                      Hydrolase         198             2674   
10                     Isomerase          20              181   
11                        Kinase         126             1994   
12                        Ligase          54              347   
13                         Lyase          40              993   
14      Membrane Traffic Protein          21               96   
15          Nucleic Acid Binding         258             2557   
16                Oxidoreductase         131              958   
17                   Phosphatase          21              144   
18                      Protease          78             1491   
19                      Receptor         178             2186   
20            Signaling Molecule         134             1236   
21          Transcription Factor         124             1521   
22      Transfer/Carrier Protein          31              842   
23                   Transferase         212             2509   
24                   Transporter          85              862   

    Population variants mapped on the structure  \
0                                        164915   
1                                          3834   
2                                          5494   
3                                          2395   
4                                          3187   
5                                         10612   
6                                          7829   
7                                         14839   
8                                          4883   
9                                         33556   
10                                         2988   
11                                        17994   
12                                         8024   
13                                         6258   
14                                         2352   
15                                        28496   
16                                        18016   
17                                         2773   
18                                        12485   
19                                        23385   
20                                        13438   
21                                         7527   
22                                         3529   
23                                        32215   
24                                        11939   

    Pathogenic variants mapped on the structure  \
0                                         32923   
1                                          1007   
2                                          2523   
3                                           472   
4                                           231   
5                                          2219   
6                                          1442   
7                                          2418   
8                                          1380   
9                                          8116   
10                                          694   
11                                         3768   
12                                          986   
13                                          885   
14                                          126   
15                                         4378   
16                                         3140   
17                                          395   
18                              


---
---
---

# [1] merge ensembl gene/protein with PANTHER annotations
- created merger of ensembl gene family description (biomart v98 with ukb ID) and PANTHER output file excluding 14 out of 3953 searched IDs

In [9]:
os.chdir("/Users/mariapalafox/Desktop/myMAP_notes/PANTHER/v14PANTHER/")
print(os.listdir())

['id3953.csv', 'PANTHERv14_ProteinRelationship_MAJORclasses.txt', 'PANTHER_MAJOR_PROCLASS_DICT.csv', '.DS_Store', 'PANTHERv14_MERGEME_3953.csv', 'PANTHERv14_3939_of_3953_UKBIDS_EverlabeledCK.txt', 'PC_major_UKBID_dict1989.csv', 'miscast_24Protein_classes.csv', 'pantherChart_3939IDs_barchart.txt', 'PANTHERv14_3939_simpledf.csv', 'CLASS_subsets_PANTHERv14', 'pantherChart_3939_barchart_protein_class.txt', 'PANTHERv14_3953_14notFount_total3939.txt', 'PANTHERv14_Protein_class_ontology_relatioship.txt', 'PANTHERv14_3939_of_3953_UKBIDS_EverlabeledCK.csv', 'panther_protein_class_value_counts_388.csv', 'MERGE_biomart_panther_proteinclass_3952Unique_ukbid.csv', 'biomart_v98_genedescription_21355.txt', 'PTHR14.1_human.txt']


In [15]:
panther3953 = pd.read_csv("PANTHERv14_MERGEME_3953.csv")
# ADDED IDs with no annotation to this file for completeness
describeMe(panther3953)

(3953, 15)
Index(['Mapped_IDs', 'GeneName_Symbol', 'PANTHER_fam_subfamily',
       'Molecular_function', 'Biological_process', 'Cellular_component',
       'Protein_class', 'Pathway', 'public_start', 'public_end', 'chr',
       'GOdb_MF_complete', 'GOdb_BP_complete', 'GOdb_CC_complete',
       'Reactome_pathway'],
      dtype='object')
Mapped_IDs                  0
GeneName_Symbol            14
PANTHER_fam_subfamily      14
Molecular_function       1666
Biological_process       1590
Cellular_component       1821
Protein_class            1964
Pathway                  3227
public_start               23
public_end                 23
chr                        23
GOdb_MF_complete          134
GOdb_BP_complete          172
GOdb_CC_complete          122
Reactome_pathway         1222
dtype: int64


In [56]:
# created biomart and panther merge file and bottom of markdown
pantherbiomart = pd.read_csv("MERGE_biomart_panther_proteinclass_3952Unique_ukbid.csv")
describeMe(pantherbiomart)

(3991, 18)
Index(['UKBID', 'GeneName_Symbol', 'PANTHER_fam_subfamily',
       'Molecular_function', 'Biological_process', 'Cellular_component',
       'Protein_class', 'Pathway', 'public_start', 'public_end', 'chr',
       'GOdb_MF_complete', 'GOdb_BP_complete', 'GOdb_CC_complete',
       'Reactome_pathway', 'Ensembl_Gene_description', 'Ensembl_Strand',
       'HGNC_symbol'],
      dtype='object')
UKBID                          0
GeneName_Symbol               14
PANTHER_fam_subfamily         14
Molecular_function          1680
Biological_process          1605
Cellular_component          1848
Protein_class               1979
Pathway                     3254
public_start                  54
public_end                    54
chr                           54
GOdb_MF_complete             134
GOdb_BP_complete             173
GOdb_CC_complete             123
Reactome_pathway            1229
Ensembl_Gene_description       0
Ensembl_Strand                 0
HGNC_symbol                    1
dtype

# Total # of specific group assignments from PANTHERv14.1

**isolated the GO (molecular funx and biological process) and PC IDs from each file saved from value coutns- to count unique number of each type of id**

- unique count values used to specify specific unique groups assigned from panther

In [26]:
proclass = panther3953['Protein_class'].value_counts()
#proclass.to_csv("panther_autoannotated_protein_class_value_counts_388groups.csv")
molfunc = panther3953['Molecular_function'].value_counts()
#molfunc.to_csv("panther_autoannotated_molecularfunx_value_counts_749groups.csv")
bioprocess = panther3953['Biological_process'].value_counts()
#bioprocess.to_csv("panther_autoannotated_bioprocess_value_counts_1179groups.csv")
# saved specific groups from panther automatic annotation in panther desktop folder

# read in files with only IDs pulled from value_counts files for molecular funx, bio process and pro class
GObio = pd.read_csv("panther_auto_GOterms_biologicalprocess_count_unique.csv", header=None)
GOmol = pd.read_csv("panther_auto_GOterms_molecularfunction_count_unique.csv", header=None)
PCcount = pd.read_csv("panther_auto_PC_count_unique.csv", header=None)

PCcount.columns = ['PCnumber']
GOmol.columns = ['GOmolfunx']
GObio.columns = ['GObioproc']
print(PCcount.head(2))
print(GOmol.head(2))
print(GObio.head(2))

uniqueCount(PCcount,'PCnumber')
print()
uniqueCount(GOmol, 'GOmolfunx')
print()
uniqueCount(GObio,'GObioproc')

  PCnumber
0  PC00202
1  PC00218
    GOmolfunx
0  GO:0003723
1  GO:0016491
    GObioproc
0  GO:0006357
1  GO:0006366
len of col:  446
len of col set:  174


len of col:  1829
len of col set:  397


len of col:  4385
len of col set:  1150



# PANTHER specifc groups 
- for IDs successfully annotated (n=1989 out of 3953)

## from value counts file, number of unique specific groups
- molecular function = 749
- biological processes = 1179
- protein class = 388

## post extracting PC/GO ids from valuecounts files, counts of unique IDs total
- molecular function = 397
- biological processes = 1150
- protein class = 174

---
---
---
---
---
---
---

In [90]:
describeMe(pantherbiomart)

(3991, 18)
Index(['UKBID', 'GeneName_Symbol', 'PANTHER_fam_subfamily',
       'Molecular_function', 'Biological_process', 'Cellular_component',
       'Protein_class', 'Pathway', 'public_start', 'public_end', 'chr',
       'GOdb_MF_complete', 'GOdb_BP_complete', 'GOdb_CC_complete',
       'Reactome_pathway', 'Ensembl_Gene_description', 'Ensembl_Strand',
       'HGNC_symbol'],
      dtype='object')
UKBID                          0
GeneName_Symbol               14
PANTHER_fam_subfamily         14
Molecular_function          1680
Biological_process          1605
Cellular_component          1848
Protein_class               1979
Pathway                     3254
public_start                  54
public_end                    54
chr                           54
GOdb_MF_complete             134
GOdb_BP_complete             173
GOdb_CC_complete             123
Reactome_pathway            1229
Ensembl_Gene_description       0
Ensembl_Strand                 0
HGNC_symbol                    1
dtype

In [91]:
pantherbiomart[pantherbiomart['HGNC_symbol'].isnull()]

UKBID                                GeneName_Symbol  \
1605  Q9NR28  Diablo homolog, mitochondrial;DIABLO;ortholog   

                              PANTHER_fam_subfamily  \
1605  DIABLO HOMOLOG, MITOCHONDRIAL (PTHR32247:SF3)   

                                     Molecular_function  \
1605  cysteine-type endopeptidase activity involved ...   

                                     Biological_process  \
1605  activation of cysteine-type endopeptidase acti...   

             Cellular_component Protein_class  \
1605  mitochondrion(GO:0005739)           NaN   

                                                Pathway  public_start  \
1605  Apoptosis signaling pathway->Second mitochondr...   122207774.0   

       public_end chr             GOdb_MF_complete  \
1605  122226052.0  12  protein binding(GO:0005515)   

                                       GOdb_BP_complete  \
1605  apoptotic process(GO:0006915);activation of cy...   

                                       GOdb_CC_complete  \
1605  mitochondrion(GO:0005739);mitochondrial interm...   

                                       Reactome_pathway  \
1605  Intrinsic Pathway for Apoptosis;Programmed Cel...   

                               Ensembl_Gene_description  Ensembl_Strand  \
1605  Diablo homolog, mitochondrial  [Source:UniProt...              -1   

     HGNC_symbol  
1605         NaN

In [95]:
len(pantherbiomart.HGNC_symbol.unique())
# duplicateRows = pantherbiomart[pantherbiomart.duplicated(['HGNC_symbol'])]
# duplicateRows
#print("Duplicate Rows based on a single column are:", duplicateRows, sep='\n')

3986

## adding protein class assignments from PANTHER (website chart view) and MISCAST (1330 genes assigned to class) to my biomart+PANTHER file with ALL UNIPROT IDs in sample (n=3953)

In [101]:
# labeling pantherbiomart merge file with miscast functional classes and PANTHER functional classes
pantherbiomart['PCmajor_PANTHER'] = pantherbiomart['UKBID']
pantherbiomart['PCmajor_MISCAST'] = pantherbiomart['HGNC_symbol']

# from dictionary made below of UKB: PC major list
ref_pan = dict(zip(groupUKB.UKBID, groupUKB.PCmajor_PANTHER))

# from dict made below of HGNC: miscast assignment
ref_mis = dict(zip(miscast_gene.HGNC_symbol, miscast_gene.PCmajor_MISCAST))

# mapping dictionaries with PANTHER and MISCAST protein class scores
pantherbiomart.PCmajor_PANTHER = pantherbiomart.PCmajor_PANTHER.map(ref_pan)
pantherbiomart.PCmajor_MISCAST = pantherbiomart.PCmajor_MISCAST.map(ref_mis)

In [102]:
pantherbiomart

UKBID                                    GeneName_Symbol  \
0     O14737     Programmed cell death protein 5;PDCD5;ortholog   
1     P62195  26S proteasome regulatory subunit 8;PSMC5;orth...   
2     Q8IYB3  Serine/arginine repetitive matrix protein 1;SR...   
3     Q9UMY4                    Sorting nexin-12;SNX12;ortholog   
4     Q12849            G-rich sequence factor 1;GRSF1;ortholog   
...      ...                                                ...   
3986  Q15121                                                NaN   
3987  Q99470                                                NaN   
3988  Q9GZP8                                                NaN   
3989  Q9NRS6                                                NaN   
3990  Q9Y3B4                                                NaN   

                                  PANTHER_fam_subfamily  \
0       PROGRAMMED CELL DEATH PROTEIN 5 (PTHR10840:SF1)   
1     26S PROTEASOME REGULATORY SUBUNIT 8 (PTHR23073...   
2     SERINE/ARGININE REPETITIVE MATRIX PROTEIN 1 (P...   
3                      SORTING NEXIN-12 (PTHR45963:SF3)   
4             G-RICH SEQUENCE FACTOR 1 (PTHR13976:SF42)   
...                                                 ...   
3986                                                NaN   
3987                                                NaN   
3988                                                NaN   
3989                                                NaN   
3990                                                NaN   

                                     Molecular_function  \
0                                                   NaN   
1                 TBP-class protein binding(GO:0017025)   
2                              mRNA binding(GO:0003729)   
3                                                   NaN   
4     nucleic acid binding(GO:0003676);structural co...   
...                                                 ...   
3986                                                NaN   
3987                                                NaN   
3988                                                NaN   
3989                                                NaN   
3990                                                NaN   

                                     Biological_process  \
0                                                   NaN   
1     RNA polymerase II transcriptional preinitiatio...   
2                           mRNA processing(GO:0006397)   
3                                                   NaN   
4            mRNA splicing, via spliceosome(GO:0000398)   
...                                                 ...   
3986                                                NaN   
3987                                                NaN   
3988                                                NaN   
3989                                                NaN   
3990                                                NaN   

                                     Cellular_component  \
0               cytosol(GO:0005829);nucleus(GO:0005634)   
1     cytosolic part(GO:0044445);nuclear lumen(GO:00...   
2                                                   NaN   
3                                                   NaN   
4                                                   NaN   
...                                                 ...   
3986                                                NaN   
3987                                                NaN   
3988                                                NaN   
3989                                                NaN   
3990                                                NaN   

                        Protein_class  \
0           enzyme modulator(PC00095)   
1                  hydrolase(PC00121)   
2     mRNA processing factor(PC00147)   
3                                 NaN   
4          ribosomal protein(PC00202)   
...                               ...   
3986                              NaN   
3987                              NaN 

In [103]:
describeMe(pantherbiomart)

(3991, 20)
Index(['UKBID', 'GeneName_Symbol', 'PANTHER_fam_subfamily',
       'Molecular_function', 'Biological_process', 'Cellular_component',
       'Protein_class', 'Pathway', 'public_start', 'public_end', 'chr',
       'GOdb_MF_complete', 'GOdb_BP_complete', 'GOdb_CC_complete',
       'Reactome_pathway', 'Ensembl_Gene_description', 'Ensembl_Strand',
       'HGNC_symbol', 'PCmajor_PANTHER', 'PCmajor_MISCAST'],
      dtype='object')
UKBID                          0
GeneName_Symbol               14
PANTHER_fam_subfamily         14
Molecular_function          1680
Biological_process          1605
Cellular_component          1848
Protein_class               1979
Pathway                     3254
public_start                  54
public_end                    54
chr                           54
GOdb_MF_complete             134
GOdb_BP_complete             173
GOdb_CC_complete             123
Reactome_pathway            1229
Ensembl_Gene_description       0
Ensembl_Strand                 0


In [105]:
# miscast 
print(3991-3492) # 499 GENES(small number redundant) have miscast annotation

print()

print(3991-1979)

499

2012


In [109]:
idcount = pantherbiomart['UKBID'].value_counts()
idcount.to_csv("MERGED_IDcounts_miscast_panther_3991.csv")

/Users/mariapalafox/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.
  


In [106]:
pantherbiomart.to_csv("MERGED_labeledPC_miscast_3991rows.csv", index=False)

In [58]:
# split into 2 files: protein class annotated and need to annotate
noclass = pantherbiomart[pantherbiomart['Protein_class'].isnull()]

withclass = panther3953[panther3953['Protein_class'].notnull()]

print(noclass.shape)
print(withclass.shape)

#noclass.to_csv("NO_MAJOR_CLASS_1964uniqueIDs.csv", index=False)

# saved and changed column order in excel to match rules for assigning PC from miscast paper
noclass = pd.read_csv("NO_MAJOR_CLASS_1964uniqueIDs.csv")
noclass

(1979, 18)
(1989, 15)


**using biomart+panther file merger for set of protein IDs with no class assigned, there are redundant ID rows in merge file because some ukb IDs have multiple genes**

## total no class protein = 1964 Uniprot IDs
## total class assigned proteins = 1989 IDs

# [2] create dictionary of MAJOR pro classes from autoannotated panther results and assign to 1989 UniProt IDs 
- using PC_all file made from merging all major protein classes from panther website bar chart to 1 file
- modified files to include PC# of major class along with ukb ID to later annotate the merger file with 3953 IDs

In [99]:
# FROM PANTHER website, used bar chart viewer to copy out all IDs within each type of major class to annotate the 1989 IDs from panther assignment
# i copied each file from PANTHER website corresponding to major protein class and added PC# to the [0] column of file
barsets = pd.read_csv("/Users/mariapalafox/Desktop/myMAP_notes/PANTHER/v14PANTHER/CLASS_subsets_PANTHERv14/PC_allclasses.txt", 
                      delimiter='\t', header=None)
bset = barsets[[0,1]].copy()
bset.columns = ['comboID', 'UKBID']
# split column create new col 
splitdf = bset['comboID'].str.split("|", n=2, expand=True)
splitdf.columns = ["PC_major", "HGNC", "UKB"]
splitdf
PCdf = pd.concat([bset, splitdf], axis=1)
PCdf.head(3)
PCdf2 = PCdf[["UKBID", "PC_major", "HGNC"]].copy()
print(len(PCdf2['UKBID'].unique())) # 1989

# ---------
def setme(x):
    return list(set(x))

PCukb = PCdf2.groupby('UKBID')['PC_major'].apply(lambda x: setme(x))
PCmajor = PCdf2.groupby('PC_major')['UKBID'].apply(lambda x: setme(x))
# -----------
groupUKB = pd.DataFrame(PCukb)
print(groupUKB.shape)
groupUKB.reset_index(inplace=True)
groupUKB.columns = ['UKBID', 'PCmajor_PANTHER']
#groupUKB.to_csv("DICT_PC_major_UKBgroup_1989.csv", index=False)
print(groupUKB.head(2))

groupPC = pd.DataFrame(PCmajor)
print(groupPC.shape)
groupPC.reset_index(inplace=True)
print(groupPC.head(2))
#groupPC.to_csv("DICT_PC_major_PCgroup_25.csv", index=False)

1989
(1989, 1)
    UKBID              PCmajor_PANTHER
0  A0AVT1           [PC00142, PC00219]
1  A1L0T0  [PC00220, PC00144, PC00176]
(25, 1)
  PC_major                                              UKBID
0  PC00060  [P31749, Q02790, P16885, Q9UPR0, P31949, Q7L5N...
1  PC00069  [Q9UNF1, P16144, P35221, Q13740, P17931, Q5EBL...


## after grouping major PC ID and UKB IDs into groupby files, assign 1989 uniprot IDs to major class

In [53]:
# 1st attempt assigning major protein class from PANTHER to simplified file
withclass = withclass[['Mapped_IDs', 'GeneName_Symbol', 'PANTHER_fam_subfamily',
       'Molecular_function', 'Biological_process', 'Cellular_component',
       'Protein_class']].copy()
withclass.reset_index(drop=True, inplace=True)


# labeling 1989 IDs with major PC from panther barchart files
groupUKB.sort_values(by=['UKBID'],inplace=True)
withclass.sort_values(by=['Mapped_IDs'],inplace=True)
withclass.columns = ['UKBID', 'GeneName_Symbol', 'PANTHER_fam_subfamily',
       'Molecular_function', 'Biological_process', 'Cellular_component',
       'Protein_class']
# merging major class assignment to 1989 UKB IDs autoannotated by panther
withclass_major = pd.merge(withclass, groupUKB, how='inner', on=['UKBID'])
withclass_major
#withclass_major.to_csv("WITH_MAJOR_CLASS_1989UKBIDs.csv", index=False)

UKBID                                    GeneName_Symbol  \
0     A0AVT1  Ubiquitin-like modifier-activating enzyme 6;UB...   
1     A1L0T0  Acetolactate synthase-like protein;ILVBL;ortholog   
2     A2A288     Probable ribonuclease ZC3H12D;ZC3H12D;ortholog   
3     A3KN83  Protein strawberry notch homolog 1;SBNO1;ortholog   
4     A5YKK6  CCR4-NOT transcription complex subunit 1;CNOT1...   
...      ...                                                ...   
1984  Q9Y6J9  TAF6-like RNA polymerase II p300/CBP-associate...   
1985  Q9Y6M9  NADH dehydrogenase [ubiquinone] 1 beta subcomp...   
1986  Q9Y6W5  Wiskott-Aldrich syndrome protein family member...   
1987  Q9Y6X8  Zinc fingers and homeoboxes protein 2;ZHX2;ort...   
1988  Q9Y6Y8         SEC23-interacting protein;SEC23IP;ortholog   

                                  PANTHER_fam_subfamily  \
0     UBIQUITIN-LIKE MODIFIER-ACTIVATING ENZYME 6 (P...   
1     ACETOLACTATE SYNTHASE-LIKE PROTEIN (PTHR18968:...   
2         RIBONUCLEASE ZC3H12D-RELATED (PTHR12876:SF11)   
3     PROTEIN STRAWBERRY NOTCH HOMOLOG 1 (PTHR12706:...   
4     CCR4-NOT TRANSCRIPTION COMPLEX SUBUNIT 1 (PTHR...   
...                                                 ...   
1984  TAF6-LIKE RNA POLYMERASE II P300/CBP-ASSOCIATE...   
1985  NADH DEHYDROGENASE [UBIQUINONE] 1 BETA SUBCOMP...   
1986  WISKOTT-ALDRICH SYNDROME PROTEIN FAMILY MEMBER...   
1987  ZINC FINGERS AND HOMEOBOXES PROTEIN 2 (PTHR154...   
1988          SEC23-INTERACTING PROTEIN (PTHR23509:SF4)   

                                     Molecular_function  \
0     ubiquitin-like modifier activating enzyme acti...   
1     lyase activity(GO:0016829);oxidoreductase acti...   
2                      nucleic acid binding(GO:0003676)   
3                                                   NaN   
4     protein-containing complex scaffold activity(G...   
...                                                 ...   
1984                                                NaN   
1985                                                NaN   
1986  actin binding(GO:0003779);structural molecule ...   
1987  DNA-binding transcription factor activity(GO:0...   
1988                 phospholipase activity(GO:0004620)   

                                     Biological_process  \
0     cellular response to DNA damage stimulus(GO:00...   
1     carbohydrate metabolic process(GO:0005975);cel...   
2                                                   NaN   
3                                                   NaN   
4     formation of translation initiation ternary co...   
...                                                 ...   
1984  RNA polymerase II transcriptional preinitiatio...   
1985                                                NaN   
1986              protein metabolic process(GO:0019538)   
1987  negative regulation of transcription by RNA po...   
1988  ER to Golgi vesicle-mediated transport(GO:0006...   

                                     Cellular_component  \
0             cytoplasm(GO:0005737);nucleus(GO:0005634)   
1                                                   NaN   
2                                                   NaN   
3                                   nucleus(GO:0005634)   
4                                    P-body(GO:0000932)   
...                                                 ...   
1984                                nucleus(GO:0005634)   
1985  mitochondrial respiratory chain complex I(GO:0...   
1986                     actin cytoskeleton(GO:0015629)   
1987                                nucleus(GO:0005634)   
1988  COPII-coated ER to Golgi transport vesicle(GO:...   

                                          Protein_class  \
0     ligase(PC00142);transfer/carrier protein(PC00219)   
1     decarboxylase(PC00089);dehydrogenase(PC00092);...   
2                         nucleic acid binding(PC00171)   
3                   helicase(PC00115);nuclease(PC00170)   
4                         transcription factor(PC00218)   
...         

## making dictionary from MISCAST protein assignment (n=1330 unique genes)
#### {GENE SYMBOL: [PC#]}

In [100]:
miscast = pd.read_csv("miscast_24Protein_classes.csv")
miscast = miscast[['PC_PANTHERCLASS', 'Protein_Class', 'Genes']].copy()
# making column into list of genes
gene2 = []
for i in miscast.Genes:
    ii = i.split(";")
    gene2.append(ii)
miscast['Genes'] = gene2 # added list column to df
# exanding list column
miscast_expand = miscast.explode('Genes')
miscast_expand.reset_index(inplace=True, drop=True)

# group by gene
def setme(x):
    return list(set(x))
miscast_gene = miscast_expand.groupby('Genes')['PC_PANTHERCLASS'].apply(lambda x: setme(x))
miscast_gene = pd.DataFrame(miscast_gene)
print(miscast_gene.shape)
miscast_gene.reset_index(inplace=True)
miscast_gene.columns = ['HGNC_symbol', 'PCmajor_MISCAST']
miscast_gene.to_csv("DICT_MISCAST_PC_GENE_1330.csv", index=False)
miscast_gene

(1330, 1)


HGNC_symbol              PCmajor_MISCAST
0            A2M  [PC00090, PC00095, PC00207]
1           AARS                    [PC00171]
2           AASS                    [PC00176]
3          ABCA1                    [PC00227]
4          ABCB6           [PC00121, PC00190]
...          ...                          ...
1325        ZEB1                    [PC00218]
1326        ZIC3                    [PC00218]
1327    ZMPSTE24           [PC00121, PC00190]
1328     ZMYND11           [PC00218, PC00171]
1329      ZNHIT3                    [PC00207]

[1330 rows x 2 columns]

In [84]:
# merging major class assignment to 1989 UKB IDs autoannotated by panther
noclass_miscast = pd.merge(noclass, miscast_gene, how='left', on=['HGNC_symbol'])
# left uses IDs from no class file 

noclass_miscast

UKBID                           Ensembl_Gene_description  \
0     Q9UMY4  sorting nexin 12 [Source:HGNC Symbol;Acc:HGNC:...   
1     Q9HBL8  NmrA like redox sensor 1 [Source:HGNC Symbol;A...   
2     Q9NU22  midasin AAA ATPase 1 [Source:HGNC Symbol;Acc:H...   
3     Q96QA5    gasdermin A [Source:HGNC Symbol;Acc:HGNC:13311]   
4     Q6VY07  phosphofurin acidic cluster sorting protein 1 ...   
...      ...                                                ...   
1974  Q15121  proliferation and apoptosis adaptor protein 15...   
1975  Q99470  stromal cell derived factor 2 [Source:HGNC Sym...   
1976  Q9GZP8  chromosome 19 open reading frame 33 [Source:HG...   
1977  Q9NRS6  sorting nexin 15 [Source:HGNC Symbol;Acc:HGNC:...   
1978  Q9Y3B4  splicing factor 3b subunit 6 [Source:HGNC Symb...   

                   Molecular_function  \
0                                 NaN   
1                                 NaN   
2                                 NaN   
3                                 NaN   
4     ion channel binding(GO:0044325)   
...                               ...   
1974                              NaN   
1975                              NaN   
1976                              NaN   
1977                              NaN   
1978                              NaN   

                                     Biological_process  \
0                                                   NaN   
1                                                   NaN   
2     rRNA processing(GO:0006364);ribosomal large su...   
3                                                   NaN   
4     protein localization to plasma membrane(GO:007...   
...                                                 ...   
1974                                                NaN   
1975                                                NaN   
1976                                                NaN   
1977                                                NaN   
1978                                                NaN   

                                       GOdb_MF_complete  \
0     protein binding(GO:0005515);enzyme binding(GO:...   
1     protein binding(GO:0005515);identical protein ...   
2     protein binding(GO:0005515);ATP binding(GO:000...   
3     phosphatidylserine binding(GO:0001786);phospha...   
4     protein binding(GO:0005515);ion channel bindin...   
...                                                 ...   
1974                                                NaN   
1975                                                NaN   
1976                                                NaN   
1977                                                NaN   
1978                                                NaN   

                                       GOdb_BP_complete HGNC_symbol  \
0     negative regulation of gene expression(GO:0010...       SNX12   
1                                urea cycle(GO:0000050)      NMRAL1   
2     ribosomal large subunit assembly(GO:0000027);r...        MDN1   
3     apoptotic process(GO:0006915);pyroptosis(GO:00...       GSDMA   
4     protein localization to Golgi apparatus(GO:003...       PACS1   
...                                                 ...         ...   
1974                                                NaN       PEA15   
1975                                                NaN        SDF2   
1976                                                NaN    C19orf33   
1977                                                NaN       SNX15   
1978                                                NaN       SF3B6   

                                        GeneName_Symbol PC_MAJOR  
0                       Sorting nexin-12;SNX12;ortholog      NaN  
1     NmrA-like family domain-containing protein 1;N...      NaN  
2                                 Midasin;MDN1;ortholog      NaN  
3                            Gasdermin-A;GSDMA;ortholog      NaN  
4     Phosphofurin acidic cluster sorting protein 1;...      NaN  
...                                 

In [85]:
describeMe(noclass_miscast)

(1979, 9)
Index(['UKBID', 'Ensembl_Gene_description', 'Molecular_function',
       'Biological_process', 'GOdb_MF_complete', 'GOdb_BP_complete',
       'HGNC_symbol', 'GeneName_Symbol', 'PC_MAJOR'],
      dtype='object')
UKBID                          0
Ensembl_Gene_description       0
Molecular_function          1322
Biological_process          1195
GOdb_MF_complete             117
GOdb_BP_complete             145
HGNC_symbol                    1
GeneName_Symbol               14
PC_MAJOR                    1770
dtype: int64


In [88]:
1770-1979 

-209

# 209 GENES were assigned major class using MISCAST, leaving 1770 genes with no class assignment now

In [89]:
noclass_miscast.to_csv("NOCLASS_miscast_assigned209_1770remaining.csv")

---
---
---

not important, making file to use as panther + ensembl annotation combo file



# CREATING BIOMART AND PANTHER FILE MERGE: 
- importing biomart v98 ensembl gene description file download

- 12.15.19
- for labeling protein class of genes that failed in PANTHER search

## come back for PDB-ENSP mapping aa positions!!!!

In [28]:
biomart = pd.read_csv("biomart_v98_genedescription_21355.txt", delimiter='\t')

biomart.columns = ['ENSG', 'ENSGv', 'Ensembl_Gene_description', 'Ensembl_Strand',
       'Ensembl_Family_Description', 'Ensembl_Protein_FamilyIDs', 'ENSP',
       'UKBID', 'PDB-ENSPmappings', 'PDB-ENSP_start', 'PDB-ENSP_end',
       'HGNC_symbol']
biomart.head(5)

nonredun_biomart = biomart[['ENSG', 'ENSGv', 'Ensembl_Gene_description', 'Ensembl_Strand',
       'Ensembl_Family_Description', 'Ensembl_Protein_FamilyIDs', 'ENSP',
       'UKBID', 'HGNC_symbol']].copy()

print(nonredun_biomart.shape) #(214277, 9)
# drop duplicates
nonredun_biomart.drop_duplicates(inplace=True)
print(nonredun_biomart.shape) #(47336, 9)

# filter biomart file for only 3953 IDs
id3953 = pd.read_csv("id3953.csv")
ls3953 = id3953.ID.tolist()
print(len(ls3953))

ENSG              ENSGv  \
0  ENSG00000198888  ENSG00000198888.2   
1  ENSG00000198888  ENSG00000198888.2   
2  ENSG00000198763  ENSG00000198763.3   
3  ENSG00000198763  ENSG00000198763.3   
4  ENSG00000198804  ENSG00000198804.2   

                            Ensembl_Gene_description  Ensembl_Strand  \
0  mitochondrially encoded NADH:ubiquinone oxidor...               1   
1  mitochondrially encoded NADH:ubiquinone oxidor...               1   
2  mitochondrially encoded NADH:ubiquinone oxidor...               1   
3  mitochondrially encoded NADH:ubiquinone oxidor...               1   
4  mitochondrially encoded cytochrome c oxidase I...               1   

                          Ensembl_Family_Description  \
0  NADH UBIQUINONE OXIDOREDUCTASE CHAIN 1 EC_7.1....   
1  NADH UBIQUINONE OXIDOREDUCTASE CHAIN 1 EC_7.1....   
2  NADH UBIQUINONE OXIDOREDUCTASE CHAIN 2 EC_7.1....   
3  NADH UBIQUINONE OXIDOREDUCTASE CHAIN 2 EC_7.1....   
4  CYTOCHROME C OXIDASE SUBUNIT 1 EC_1.9.3.1 CYTO...   

  Ensembl_Protein_FamilyIDs             ENSP   UKBID PDB-ENSPmappings  \
0             PTHR11432_SF3  ENSP00000354687  P03886           5xtc.s   
1             PTHR11432_SF3  ENSP00000354687  P03886           5xtd.s   
2            PTHR22773_SF41  ENSP00000355046  P03891           5xtc.i   
3            PTHR22773_SF41  ENSP00000355046  P03891           5xtd.i   
4                 PTHR10422  ENSP00000354499  P00395           5z62.A   

   PDB-ENSP_start  PDB-ENSP_end HGNC_symbol  
0             1.0         318.0      MT-ND1  
1             1.0         318.0      MT-ND1  
2             1.0         347.0      MT-ND2  
3             1.0         347.0      MT-ND2  
4             1.0         513.0      MT-CO1

In [44]:
biomart2 = addcolumnconditionalDrop(ls3953, nonredun_biomart, 'UKBID', 'in3953')
biomart2.shape

biomart3 = biomart2[['Ensembl_Gene_description', 'Ensembl_Strand',
       'UKBID', 'HGNC_symbol']].copy()
biomart3.drop_duplicates(inplace=True)
biomart3.shape #(3991, 4)

  in3953  Count
0  False  36823
1   True  10513
dropping false
df shape post drop:  (10513, 10)


(10513, 10)

In [83]:
checkColumnValues(biomart3, "UKBID")

       UKBID  Count
0     P62805     14
1     P68431     10
2     P0C0S8      5
3     P23610      3
4     Q99666      2
...      ...    ...
3947  P22307      1
3948  Q16775      1
3949  Q14320      1
3950  Q9NP97      1
3951  Q9H204      1

[3952 rows x 2 columns]


### reason why there are more rows that uniprot IDs in ensembl file
- 1 uniprot ID has multiple gene symbols 

In [84]:
biomart3[biomart3['UKBID'] == 'P62805']

Ensembl_Gene_description  Ensembl_Strand  \
148   histone cluster 1 H4 family member k [Source:H...              -1   
155   histone cluster 1 H4 family member j [Source:H...               1   
157   histone cluster 1 H4 family member l [Source:H...              -1   
3724  histone cluster 1 H4 family member a [Source:H...               1   
3726  histone cluster 1 H4 family member b [Source:H...              -1   
3792  histone cluster 1 H4 family member c [Source:H...               1   
4992  histone cluster 1 H4 family member i [Source:H...               1   
5359  histone cluster 1 H4 family member d [Source:H...              -1   
5381  histone cluster 1 H4 family member e [Source:H...               1   
5419  histone cluster 1 H4 family member f [Source:H...               1   
5442  histone cluster 1 H4 family member h [Source:H...              -1   
7252  histone cluster 2 H4 family member a [Source:H...               1   
7264  histone cluster 2 H4 family member b [Source:H...              -1   
8033  histone cluster 4 H4 [Source:HGNC Symbol;Acc:H...              -1   

       UKBID HGNC_symbol  
148   P62805    HIST1H4K  
155   P62805    HIST1H4J  
157   P62805    HIST1H4L  
3724  P62805    HIST1H4A  
3726  P62805    HIST1H4B  
3792  P62805    HIST1H4C  
4992  P62805    HIST1H4I  
5359  P62805    HIST1H4D  
5381  P62805    HIST1H4E  
5419  P62805    HIST1H4F  
5442  P62805    HIST1H4H  
7252  P62805    HIST2H4A  
7264  P62805    HIST2H4B  
8033  P62805     HIST4H4

In [89]:
# MERGE file created and imported above for panther protein class 
MER = pd.merge(panther3953, biomart3, how='inner', on=['UKBID'])
len(MER.UKBID.unique()) #3952
MER.to_csv("MERGE_biomart_panther_proteinclass_3952Unique_ukbid.csv", index=False)
MER

UKBID                                    GeneName_Symbol  \
0     O14737     Programmed cell death protein 5;PDCD5;ortholog   
1     P62195  26S proteasome regulatory subunit 8;PSMC5;orth...   
2     Q8IYB3  Serine/arginine repetitive matrix protein 1;SR...   
3     Q9UMY4                    Sorting nexin-12;SNX12;ortholog   
4     Q12849            G-rich sequence factor 1;GRSF1;ortholog   
...      ...                                                ...   
3986  Q15121                                                NaN   
3987  Q99470                                                NaN   
3988  Q9GZP8                                                NaN   
3989  Q9NRS6                                                NaN   
3990  Q9Y3B4                                                NaN   

                                  PANTHER_fam_subfamily  \
0       PROGRAMMED CELL DEATH PROTEIN 5 (PTHR10840:SF1)   
1     26S PROTEASOME REGULATORY SUBUNIT 8 (PTHR23073...   
2     SERINE/ARGININE REPETITIVE MATRIX PROTEIN 1 (P...   
3                      SORTING NEXIN-12 (PTHR45963:SF3)   
4             G-RICH SEQUENCE FACTOR 1 (PTHR13976:SF42)   
...                                                 ...   
3986                                                NaN   
3987                                                NaN   
3988                                                NaN   
3989                                                NaN   
3990                                                NaN   

                                     Molecular_function  \
0                                                   NaN   
1                 TBP-class protein binding(GO:0017025)   
2                              mRNA binding(GO:0003729)   
3                                                   NaN   
4     nucleic acid binding(GO:0003676);structural co...   
...                                                 ...   
3986                                                NaN   
3987                                                NaN   
3988                                                NaN   
3989                                                NaN   
3990                                                NaN   

                                     Biological_process  \
0                                                   NaN   
1     RNA polymerase II transcriptional preinitiatio...   
2                           mRNA processing(GO:0006397)   
3                                                   NaN   
4            mRNA splicing, via spliceosome(GO:0000398)   
...                                                 ...   
3986                                                NaN   
3987                                                NaN   
3988                                                NaN   
3989                                                NaN   
3990                                                NaN   

                                     Cellular_component  \
0               cytosol(GO:0005829);nucleus(GO:0005634)   
1     cytosolic part(GO:0044445);nuclear lumen(GO:00...   
2                                                   NaN   
3                                                   NaN   
4                                                   NaN   
...                                                 ...   
3986                                                NaN   
3987                                                NaN   
3988                                                NaN   
3989                                                NaN   
3990                                                NaN   

                        Protein_class  \
0           enzyme modulator(PC00095)   
1                  hydrolase(PC00121)   
2     mRNA processing factor(PC00147)   
3                                 NaN   
4          ribosomal protein(PC00202)   
...                               ...   
3986                              NaN   
3987                              NaN 